File: ZuluTimeGPS/ZuluTimeGPS/Services/LocationService.swift
import Foundation
import CoreLocation
import Combine

class LocationService: NSObject, ObservableObject, CLLocationManagerDelegate {
    @Published private(set) var currentLocation: CLLocation?
    @Published private(set) var statusMessage = "Location not initialized"
    @Published private(set) var isTracking = false

    private let locationManager = CLLocationManager()

    // Formatted coordinate strings
    var latitude: String {
        guard let loc = currentLocation else { return "--" }
        return String(format: "%.6f", loc.coordinate.latitude)
    }

    var longitude: String {
        guard let loc = currentLocation else { return "--" }
        return String(format: "%.6f", loc.coordinate.longitude)
    }

    var altitude: String {
        guard let loc = currentLocation else { return "--" }
        return String(format: "%.1f m", loc.altitude)
    }

    var accuracy: String {
        guard let loc = currentLocation else { return "--" }
        return String(format: "\u{00B1}%.1f m", loc.horizontalAccuracy)
    }

    var capGrid: String {
        guard let loc = currentLocation else { return "--" }
        return CAPGridService.gridReference(for: loc.coordinate) ?? "Outside coverage"
    }

    override init() {
        super.init()
        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
        locationManager.distanceFilter = 10 // Update every 10 meters
        locationManager.allowsBackgroundLocationUpdates = true
        locationManager.pausesLocationUpdatesAutomatically = false
    }

    // MARK: - Public Methods

    func startTracking() {
        guard !isTracking else { return }

        let status = locationManager.authorizationStatus
        switch status {
        case .notDetermined:
            statusMessage = "Requesting permission..."
            locationManager.requestWhenInUseAuthorization()
        case .restricted, .denied:
            statusMessage = "Location permission denied"
        case .authorizedWhenInUse, .authorizedAlways:
            beginLocationUpdates()
        @unknown default:
            statusMessage = "Unknown authorization status"
        }
    }

    func stopTracking() {
        locationManager.stopUpdatingLocation()
        isTracking = false
        statusMessage = "Tracking stopped"
    }

    func refreshLocation() {
        let status = locationManager.authorizationStatus
        guard status == .authorizedWhenInUse || status == .authorizedAlways else {
            statusMessage = "Location permission not granted"
            return
        }
        statusMessage = "Getting location..."
        locationManager.requestLocation()
    }

    // MARK: - Private Methods

    private func beginLocationUpdates() {
        isTracking = true
        statusMessage = "Acquiring location..."
        locationManager.startUpdatingLocation()
    }

    // MARK: - CLLocationManagerDelegate

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }
        currentLocation = location
        statusMessage = isTracking ? "Location Active" : "Location updated"
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        statusMessage = "Error: \(error.localizedDescription)"
    }

    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        switch manager.authorizationStatus {
        case .authorizedWhenInUse, .authorizedAlways:
            statusMessage = "Location permission granted"
            beginLocationUpdates()
        case .denied:
            statusMessage = "Location permission denied"
            isTracking = false
        case .restricted:
            statusMessage = "Location access restricted"
            isTracking = false
        case .notDetermined:
            break
        @unknown default:
            break
        }
    }
}

File: zulutime_gps/lib/services/location_service.dart
import 'package:flutter/foundation.dart';
import 'package:geolocator/geolocator.dart';
import 'package:permission_handler/permission_handler.dart';

class LocationService with ChangeNotifier {
  Position? _currentPosition;
  String _statusMessage = 'Location not initialized';
  bool _isTracking = false;
  
  Position? get currentPosition => _currentPosition;
  String get statusMessage => _statusMessage;
  bool get isTracking => _isTracking;
  
  // Formatted coordinates
  String get latitude => _currentPosition != null 
      ? _currentPosition!.latitude.toStringAsFixed(6) 
      : '--';
  
  String get longitude => _currentPosition != null 
      ? _currentPosition!.longitude.toStringAsFixed(6) 
      : '--';
  
  String get altitude => _currentPosition != null 
      ? '${_currentPosition!.altitude.toStringAsFixed(1)} m' 
      : '--';
  
  String get accuracy => _currentPosition != null 
      ? '±${_currentPosition!.accuracy.toStringAsFixed(1)} m' 
      : '--';
  
  // Check and request location permissions
  Future<bool> requestLocationPermission() async {
    // Check if location services are enabled
    bool serviceEnabled = await Geolocator.isLocationServiceEnabled();
    if (!serviceEnabled) {
      _statusMessage = 'Location services disabled';
      notifyListeners();
      return false;
    }
    
    // Request permission
    var status = await Permission.location.status;
    
    if (status.isDenied) {
      status = await Permission.location.request();
    }
    
    if (status.isPermanentlyDenied) {
      _statusMessage = 'Location permission permanently denied';
      notifyListeners();
      return false;
    }
    
    if (status.isGranted) {
      _statusMessage = 'Location permission granted';
      notifyListeners();
      return true;
    }
    
    _statusMessage = 'Location permission denied';
    notifyListeners();
    return false;
  }
  
  // Start location tracking
  Future<void> startTracking() async {
    if (_isTracking) return;
    
    bool hasPermission = await requestLocationPermission();
    if (!hasPermission) return;
    
    _isTracking = true;
    _statusMessage = 'Acquiring location...';
    notifyListeners();
    
    try {
      // Get initial position
      _currentPosition = await Geolocator.getCurrentPosition(
        desiredAccuracy: LocationAccuracy.high,
      );
      _statusMessage = 'Location acquired';
      notifyListeners();
      
      // Start continuous tracking
      const LocationSettings locationSettings = LocationSettings(
        accuracy: LocationAccuracy.high,
        distanceFilter: 10, // Update every 10 meters
      );
      
      Geolocator.getPositionStream(locationSettings: locationSettings).listen(
        (Position position) {
          _currentPosition = position;
          _statusMessage = 'Tracking active';
          notifyListeners();
        },
        onError: (error) {
          _statusMessage = 'Error: ${error.toString()}';
          notifyListeners();
        },
      );
    } catch (e) {
      _statusMessage = 'Error: ${e.toString()}';
      _isTracking = false;
      notifyListeners();
    }
  }
  
  // Stop location tracking
  void stopTracking() {
    _isTracking = false;
    _statusMessage = 'Tracking stopped';
    notifyListeners();
  }
  
  // Get one-time location
  Future<void> getCurrentLocation() async {
    bool hasPermission = await requestLocationPermission();
    if (!hasPermission) return;
    
    try {
      _statusMessage = 'Getting location...';
      notifyListeners();
      
      _currentPosition = await Geolocator.getCurrentPosition(
        desiredAccuracy: LocationAccuracy.high,
      );
      _statusMessage = 'Location updated';
      notifyListeners();
    } catch (e) {
      _statusMessage = 'Error: ${e.toString()}';
      notifyListeners();
    }
  }
}
